### PR TITLE
Issue 56: Add tool for extracting packets from PCAP file

### LINKED ISSUE #56: Tool for extracting packet bytes from PCAP file
The tool should be able to extract all packets of a specified protocol layer from a PCAP file and write the byte representation of each packet into a separate file.

### PR DESCRIPTION
**Pull request by treiher**
# Description

Closes #56

# Review Checklist

### Implementation

- [x] Does this code change accomplish what it is supposed to do?
- [x] Can this solution be simplified?
- [x] Does this change add unwanted compile-time or run-time dependencies?
- [x] Was a framework, API, library, service used that should not be used?
- [x] Was a framework, API, library, service not used that could improve the solution?
- [x] Is the code at the right abstraction level?
- [x] Is the code modular enough?
- [x] Would you have solved the problem in a different way that is substantially better in terms of the code’s maintainability, readability, performance, security?
- [x] Does similar functionality already exist in the codebase? If so, why isn’t this functionality reused?
- [x] Are there any best practices, design patterns or language-specific patterns that could substantially improve this code?

### Logic Errors and Bugs

- [x] Can you think of any use case in which the code does not behave as intended?
- [x] Can you think of any inputs or external events that could break the code?

### Error Handling and Logging

- [x] Is error handling done the correct way?
- [x] Should any logging or debugging information be added or removed?
- [x] Are error messages user-friendly?
- [x] Are there enough log events and are they written in a way that allows for easy debugging?
- [x] Is the style of error / logging messages consistent?

### Dependencies

- [x] If this change requires updates outside of the code, like updating the documentation, configuration, readme files, was this done?
- [x] Might this change have any ramifications for other parts of the system, or backward compatibility?

### Security and Data Privacy

- [x] Does this code open the software up for security vulnerabilities?
- [x] Does this code change reveal some secret information like keys, passwords, or usernames (e.g., checked-in credentials)?

### Performance

- [x] Do you think this code change will impact system performance in a negative way?
- [x] Do you see any potential to improve the performance of the code?

### Usability

- [x] Is the proposed solution well designed from a usability perspective?
- [x] Is the API well documented?
- [x] Is the API/UI intuitive to use?

### Testing

- [x] Does it have enough automated tests (unit/integration/system tests)?
- [x] Are there some test cases, input or edge cases that should be tested in addition?

### Readability

- [x] Was the code easy to understand?
- [x] Can the readability of the code be improved by smaller methods?
- [x] Can the readability of the code be improved by different function/method or variable names?
- [x] Is the code located in the right file/folder/package?
- [x] Do you think certain methods should be restructured to have a more intuitive control flow?
- [x] Is the data flow understandable?
- [x] Are there redundant comments?
- [x] Could some comments convey the message better?
- [x] Would additional comments make the code more understandable?
- [x] Could some comments be removed by making the code itself more readable?
- [x] Is there any commented out code?

### Experts Opinion

- [x] Do you think a specific expert, like a security expert or a usability expert, should look over the code before it can be committed?
- [x] Will this code change impact different teams?
- [x] Should they have a say on the change as well?

### CODE REVIEW

File: tools/extract_packets.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env python3
```
Comment (senier) [inline tools/extract_packets.py:1]: I wonder whether we should add a `requirements.txt` in `tools/` (especially now when we use scapy).

File: tools/extract_packets.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+"""
+This tool extracts all packets of a specified protocol layer from a PCAP file and writes the byte
+representation of each packet into a separate file.
```
Comment (senier) [inline tools/extract_packets.py:5]: The tool in its current form has not been very useful for me. I tried to extract a couple of protocols I needed recently:

- PPPoE (seems to be unsupported by scapy)
- HTTP/2 (seems to be unsupported by scapy)
- IPv6 (I used the "inet6" layer, but nothing was dumped)

I have the feeling that we won't get very far with Scapy. Probably something based on Wireshark would have better chances of supporting protocols we are interested in.

In general I wonder whether it would be better to extract the payload of an upper layer protocol. This would allow us to request the payload field of Ethernet to be extracted whenever Ether_Type is 0x86dd when we want IPv6 packets.

**Review by senier — CHANGES_REQUESTED** [review]

File: tools/extract_packets.py (hunk 2 above)
Comment (treiher) [reply tools/extract_packets.py:5]: >     * IPv6 (I used the "inet6" layer, but nothing was dumped)

A specific class in a Scapy layer package has to be given:

```
usage: extract_packets.py [-h] LAYER PCAP_FILE OUTPUT_DIRECTORY

positional arguments:
  LAYER             Scapy layer in the form <package>.<class> (e.g., dhcp.BOOTP)
  PCAP_FILE
  OUTPUT_DIRECTORY
```

For IPv6 the layer argument must be `inet6.IPv6`.

> I have the feeling that we won't get very far with Scapy. Probably something based on Wireshark would have better chances of supporting protocols we are interested in.

Creating a tool based on Scapy was a very easy and fast solution for me and it worked very well for DHCP. There is still a number of [protocols](https://scapy.readthedocs.io/en/latest/api/scapy.layers.html) for which it could be useful for us in the future. Nevertheless, you are probably right and looking into a more powerful alternative would make sense.

> In general I wonder whether it would be better to extract the payload of an upper layer protocol. This would allow us to request the payload field of Ethernet to be extracted whenever Ether_Type is 0x86dd when we want IPv6 packets.

I think the current approach works fine for all layers directly supported by Scapy, but maybe your described approach could be helpful to extract unsupported layers as well.

File: tools/extract_packets.py (hunk 1 above)
Comment (treiher) [reply tools/extract_packets.py:1]: I added a separate `requirements.txt` in fa57e5d.

File: tools/extract_packets.py (hunk 2 above)
Comment (treiher) [reply tools/extract_packets.py:5]: I have improved the argument validation in c089c86. `inet6` is detected as an invalid layer argument now.

File: tools/extract_packets.py (hunk 1 above)
Comment (treiher) [reply tools/extract_packets.py:1]: The pylint check will fail when Scapy is not installed:

https://github.com/Componolit/RecordFlux-specifications/runs/2110974130?check_suite_focus=true

IMHO all checks and tests should run successfully after installing all dependencies defined in the `requirements.txt` in the top-level directory. I therefore tend to revert this change.

Comment (treiher) [reply tools/extract_packets.py:1]: Now there is a single `requirements.txt` in the top-level directory again.

File: tools/extract_packets.py (hunk 2 above)
Comment (treiher) [reply tools/extract_packets.py:5]: I have added an option to show a list of all available Scapy layers in 1fed3a3.

File: tools/extract_packets.py (hunk 1 above)
Comment (senier) [reply tools/extract_packets.py:1]: Sorry for the confusion.

File: tools/extract_packets.py (hunk 2 above)
Comment (senier) [reply tools/extract_packets.py:5]: That option is very helpful.

**Review by senier — APPROVED** [review]

**Review by senier — APPROVED** [review]
